src/ServicePulse.Host/vue/src/components/failedmessages/FlowDiagram.vue

```diff
@@ -61,292 +58,176 @@ function mapMessage(message) {
+            //if the current parent node is the same as the previous parent node, then the current position needs to be to the right of siblings
+            const currentParentWidth = previousParent === parentMessage ? currentWidth : 0;
+            const startX = parentMessage == null ? 0 : parentMessage.XPos - parentMessage.width / 2;
+            //store the position of the node against the message, so child nodes can use it to determine their start position
+            message.XPos = startX + (currentParentWidth + message.width / 2);
+            return {
+              result: [
+                ...result,
+                {
+                  id: `${message.messageId}##${message.receivingEndpoint}`,
+                  type: "message",
+                  data: message,
+                  label: message.nodeName,
+                  position: { x: message.XPos * 300, y: message.level * 200 },
+                },
+              ],
+              currentWidth: currentParentWidth + message.width,
+              previousParent: parentMessage,
+            };
+          },
+          { result: [], currentWidth: 0, previousParent: null }
+        ).result;
+      })
+  );
+}
+
+function constructEdges(mappedMessages) {
+  return mappedMessages
+    .filter((message) => message.parentId)
+    .map((message) => ({
+      id: `${message.parentId}##${message.messageId}`,
+      source: `${message.parentId}##${message.parentEndpoint}`,
+      target: `${message.messageId}##${message.receivingEndpoint}`,
+      markerEnd: MarkerType.ArrowClosed,
+      style: {
+        "stroke-dasharray": message.type === "Event message" && "5, 3",
+      },
+    }));
+}
+
+const elements = ref([]);
+const { onPaneReady, fitView } = useVueFlow();
+
+onMounted(async () => {
+  const messages = await getConversation(props.conversationId);
+  const mappedMessages = messages.map(mapMessage);
+
+  const assignDescendantLevelsAndWidth = (message, level = 0) => {
+    message.level = level;
+    const children = mappedMessages.filter((mm) => mm.parentId === message.messageId && mm.parentEndpoint === message.receivingEndpoint);
+    message.width =
+      children.length === 0
+        ? 1 //leaf node
+        : children.map((child) => (child.width == null ? assignDescendantLevelsAndWidth(child, level + 1) : child)).reduce((sum, { width }) => sum + width, 0);
+    return message;
+  };
+  for (const root of mappedMessages.filter((message) => !message.parentId)) assignDescendantLevelsAndWidth(root);
 
-  for (i = 0; i < messages.length; i += 1) {
-    map[messages[i].messageId] = i; // initialize the map
-    messages[i].children = []; // initialize the children
-  }
+  elements.value = [...constructNodes(mappedMessages), ...constructEdges(mappedMessages)];
+  //TODO: if doing fitView on next Vue onUpdated() then it doesn't appear the elements
+  // are actually drawn yet. See if we can determine a better event to use this on rather than relying on setTimeout
```

Comment:
Still working on this one?

Reply:
I couldn't work out a nice solution (the documentation for Vue-flow isn't the greatest). This was the best I could come up with, but left the TODO for future reference